src/snowflake/snowpark/dataframe.py

```diff
@@ -2791,35 +2834,71 @@ def union_all_by_name(
             self._set_ast_ref(ast.df)
             other._set_ast_ref(ast.other)
 
-        return self._union_by_name_internal(other, is_all=True, _ast_stmt=stmt)
+        return self._union_by_name_internal(
+            other,
+            is_all=True,
+            allow_missing_columns=allow_missing_columns,
+            _ast_stmt=stmt,
+        )
 
     def _union_by_name_internal(
-        self, other: "DataFrame", is_all: bool = False, _ast_stmt: proto.Assign = None
+        self,
+        other: "DataFrame",
+        is_all: bool = False,
+        allow_missing_columns: bool = False,
+        _ast_stmt: proto.Assign = None,
     ) -> "DataFrame":
-        left_output_attrs = self._output
-        right_output_attrs = other._output
-        right_output_attr_by_name = {rattr.name: rattr for rattr in right_output_attrs}
-
-        try:
-            right_project_list = [
-                right_output_attr_by_name[lattr.name] for lattr in left_output_attrs
-            ]
-        except KeyError:
-            missing_lattrs = [
-                lattr.name
-                for lattr in left_output_attrs
-                if lattr.name not in right_output_attr_by_name
-            ]
-            raise SnowparkClientExceptionMessages.DF_CANNOT_RESOLVE_COLUMN_NAME_AMONG(
-                ", ".join(missing_lattrs),
-                ", ".join(list(right_output_attr_by_name.keys())),
-            )
-
-        not_found_attrs = [
-            rattr for rattr in right_output_attrs if rattr not in right_project_list
-        ]
+        left_cols = {attr.name for attr in self._output}
+        left_attr_map = {attr.name: attr for attr in self._output}
+        right_cols = {attr.name for attr in other._output}
+        right_attr_map = {attr.name: attr for attr in other._output}
+
+        missing_left = right_cols - left_cols
+        missing_right = left_cols - right_cols
+
+        if missing_left or missing_right:
+            if allow_missing_columns:
+                left = self
+                right = other
+                if missing_left:
+                    # schema has to be resolved in order to get correct type information for missing columns
+                    dt_map = {
+                        field.name: field.datatype for field in right.schema.fields
+                    }
+                    # depending on how column names are handled names may differ from attributes to schema
+                    materialized_names = {
+                        StructField(col, DataType()).name for col in missing_left
+                    }
+                    left = left.select(
+                        "*",
+                        *[
+                            lit(None).cast(dt_map[col]).alias(col)
+                            for col in materialized_names
+                        ],
+                    )
```

Comment:
nit: can we make this a reusable function? handling left and right seems identical